relay/DCUtR.md

```diff
@@ -0,0 +1,179 @@
+connectivity with hole punching that eschews signaling servers and utilizes
+existing relay connections instead. That is, peers start with a relay connection
+and synchronize directly, without the use of a signaling server. If the hole
+punching attempt is successful, the peers _upgrade_ their connection to a direct
+connection and they can close the relay connection. If the hole punching attempt
+fails, they can keep using the relay connection as they were.
+
+## The Protocol
+
+Consider two peers, `A` and `B`. `A` wants to connect to `B`, which is
+behind a NAT and advertises relay addresses. `A` may itself be behind
+a NAT or be a public node.
+
+The protocol starts with the completion of a relay connection from `A`
+to `B`.  Upon observing the new connection, the inbound peer (here `B`)
+checks the addresses advertised by `A` via identify. If that set
+includes public addresses, then `A` _may_ be reachable by a direct
+connection, in which case `B` attempts a unilateral connection upgrade
+by initiating a direct connection to `A`.
+
+If the unilateral connection upgrade attempt fails or if `A` is itself a NATed
+peer that doesn't advertise public address, then `B` initiates the direct
+connection upgrade protocol as follows:
+1. `B` opens a stream to `A` using the `/libp2p/dcutr` protocol.
+2. `B` sends to `A` a `Connect` message containing its observed (and possibly
+   predicted) addresses from identify and starts a timer to measure RTT of the
+   relay connection.
+3. Upon receving the `Connect`, `A` responds back with a `Connect` message
+   containing its observed (and possibly predicted) addresses.
+4. Upon receiving the `Connect`, `B` sends a `Sync` message and starts a timer
+   for half the RTT measured from the time between sending the initial `Connect`
+   and receiving the response.
+5. Simultaneous Connect. This depends on the transport in use:
+   - For TCP:
+      - Upon receiving the `Sync`, `A` immediately attempts to directly connect
+        to `B`. `A` does so by dialling the addresses obtained from the
+        `Connect` message in parallel.
+      - Upon expiry of the timer, `B` attempts to directly connect to `A`. `B`
+        does so by dialing the addresses obtained from the `Connect` message in
+        parallel.
+      - This will result in a TCP Simultaneous Connect. For the purpose of all
+        protocols run on top of this TCP connection, `A` is assumed to be the
+        client and `B` the server.
+   - For QUIC:
+      - Upon receiving the `Sync`, `A` immediately attempts to directly connect
+        to `B`. `A` does so by dialing the addresses obtained from the `Connect`
+        message in parallel.
+      - Upon expiry of the timer, `B` starts to send UDP packets filled with
+        random bytes to the addresses obtained from the `Connect` message.
+        Packets should be sent in random intervals between 10 and 200 ms to each
+        address.
```

Comment:
I agree with the intention, but this might be a bit confusing, since `Connect` could also contain TCP addresses.

Maybe we should prefix the entire section (5.) with a "For every address obtained from the `Connect` message, ...", and then clean up the subpoints?

Reply:
Done via https://github.com/libp2p/specs/pull/173/commits/6f558f13aa4cbfe8458a3207764bd52e3f61a573.